### ISSUE TITLE
Location of LD_FLAGS causes the C compiler on z/OS grief

### ISSUE THREAD

**Issue by MikeFultonDev** [madler/zlib#735]
In the Makefile.in, there are a number of instances where the linker options are _after_ the object file. This works with most compilers, but not the C compiler on z/OS, which complains. Can the order of options be switched? The change I had to make is as follows:

```
 example$(EXE): example.o $(STATICLIB)
- $(CC) $(CFLAGS) -o $@ example.o $(TEST_LDFLAGS)
+ $(CC) $(CFLAGS) -o $@ $(TEST_LDFLAGS) example.o

 minigzip$(EXE): minigzip.o $(STATICLIB)
- $(CC) $(CFLAGS) -o $@ minigzip.o $(TEST_LDFLAGS)
+ $(CC) $(CFLAGS) -o $@ $(TEST_LDFLAGS) minigzip.o

 examplesh$(EXE): example.o $(SHAREDLIBV)
- $(CC) $(CFLAGS) -o $@ example.o -L. $(SHAREDLIBV)
+ $(CC) $(CFLAGS) -L. -o $@ example.o $(SHAREDLIBV)

 minigzipsh$(EXE): minigzip.o $(SHAREDLIBV)
- $(CC) $(CFLAGS) -o $@ minigzip.o -L. $(SHAREDLIBV)
+ $(CC) $(CFLAGS) -L. -o $@ minigzip.o $(SHAREDLIBV)

 example64$(EXE): example64.o $(STATICLIB)
- $(CC) $(CFLAGS) -o $@ example64.o $(TEST_LDFLAGS)
+ $(CC) $(CFLAGS) -o $@ $(TEST_LDFLAGS) example64.o

 minigzip64$(EXE): minigzip64.o $(STATICLIB)
- $(CC) $(CFLAGS) -o $@ minigzip64.o $(TEST_LDFLAGS)
+ $(CC) $(CFLAGS) -o $@ $(TEST_LDFLAGS) minigzip64.o
```

**Comment by thesamesam** [comment]
LDFLAGS always needs to be before object files to ensure they work correctly anyway. Not that it's relevant in this case, but for example `-Wl,--as-needed` requires it.

**Comment by MikeFultonDev** [comment]
@thesamesam ah. so you are saying that I should have the ${SHAREDLIBV} before the .o's ? But the other lines are ok?

**Comment by madler** [comment]
Applied.

**Comment by madler** [comment]
Reverted. This broke other builds. Sorry.

**Comment by Neustradamus** [comment]
@MikeFultonDev, @thesamesam: Can you look to have a good solution?

**Comment by MikeFultonDev** [comment]
@madler do you have details on what broke? @IgorTorodovskiIBM and @perry-ca any suggestions given the push to clang on z/OS now (are there alternatives now?)

**Comment by perry-ca** [comment]
I see in the Makefile.in
```
TEST_LDFLAGS=$(LDFLAGS) -L. libz.a
```
This is causing the dilemma you are seeing.  The `$(LDFLAGS)` should go before the object files and `-L libz.a` should go after the object files.  I suggest separating this into two separate variables:
```
TEST_LDFLAGS=$(LDFLAGS)
TEST_LIBS=-L libz.a
```

Put `TEST_LDFLAGS` before the .o files and `TEST_LIBS` after.

**Comment by Neustradamus** [comment]
@MikeFultonDev: Have you found a solution?

**Comment by MikeFultonDev** [comment]
Let me try again - we have switched to a clang compiler from our older compiler, so I'm hoping this problem has gone away.

**Comment by MikeFultonDev** [comment]
Confirmed. With the clang compiler, we no longer need a patch. Apologies for not updating this earlier.

**Comment by Neustradamus** [comment]
@MikeFultonDev: Thanks for your answers and your solution :)
But if people do not use clang compiler?

**Comment by madler** [comment]
I made the change recommended by @perry-ca , which appears to cause no harm. https://github.com/madler/zlib/commit/1a8db63788c34a50e39e273d39b7e1033208aea2

**Comment by perry-ca** [comment]
> I made the change recommended by @perry-ca , which appears to cause no harm. [1a8db63](https://github.com/madler/zlib/commit/1a8db63788c34a50e39e273d39b7e1033208aea2)

Did it fix the problem?
